Client/Frontend/Browser/Favicons/FaviconFetcher.swift

```diff
@@ -0,0 +1,530 @@
+                    var path = $0.1["image_url"].string else {
+                        return
+                }
+                path = path.replacingOccurrences(of: ".png", with: "")
+                let filePath = Bundle.main.path(forResource: "TopSites/" + path, ofType: "png")
+                if let filePath = filePath {
+                    if color == "#fff" {
+                        icons[url] = (UIColor.white, filePath)
+                    } else {
+                        icons[url] = (UIColor(colorString: color.replacingOccurrences(of: "#", with: "")), filePath)
+                    }
+                }
+            })
+            return icons
+        } catch {
+            log.error("Failed to get default icons at \(filePath): \(error.localizedDescription)")
+            return [:]
+        }
+    }()
+     
+    // MARK: - Fetched Icons
+    
+    private func downloadIcon(url: URL, addingToDatabase: Bool, completion: @escaping (UIImage?) -> Void) {
+        // Fetch favicon directly
+        var imageOperation: SDWebImageOperation?
+        
+        let onProgress: ImageCacheProgress = { receivedSize, expectedSize, _ in
+            if receivedSize > FaviconHandler.maximumFaviconSize || expectedSize > FaviconHandler.maximumFaviconSize {
+                imageOperation?.cancel()
+            }
+        }
+        
+        let onCompletion: ImageCacheCompletion = { [weak self] image, _, _, cacheType, url in
+            guard let self = self else { return }
+            let favicon = Favicon(url: url.absoluteString, date: Date(), type: self.kind.iconType)
+            
+            if let image = image {
+                favicon.width = Int(image.size.width)
+                favicon.height = Int(image.size.height)
+                if addingToDatabase && !PrivateBrowsingManager.shared.isPrivateBrowsing {
+                    FaviconMO.add(favicon, forSiteUrl: self.url)
+                }
+                
+                completion(image)
+            } else {
+                favicon.width = 0
+                favicon.height = 0
+                
+                completion(nil)
+            }
+        }
+        
+        imageOperation = WebImageCacheWithNoPrivacyProtectionManager.shared.load(from: url, progress: onProgress, completion: onCompletion)
+        if let op = imageOperation {
+            imageOps.append(op)
+        }
+    }
+    
+    private func fetchIcon(_ completion: @escaping (URL, FaviconAttributes) -> Void) {
+        // Attempt to find favicon cached for the given Domain
+        if let favicon = domain.favicon, let urlString = favicon.url, let url = URL(string: urlString) {
+            // Verify that the favicon we have on file is what we want to pull
+            // If not, we will just default to monogram to avoid blurry images
+            if faviconOnFileMatchesFetchKind(favicon) {
+                downloadIcon(url: url, addingToDatabase: false) { [weak self] image in
```

Comment:
Setting it to false always is triggering multiple requests for sites like `coinmarketcap`

Reply:
That doesnt make sense, if we've reached this point it means we've pulled an icon from FaviconMO already (meaning its already in the DB and doesn't need to be added)